parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/ProcessingShard.java

```diff
@@ -182,4 +188,15 @@ private static String cantTakeAsWorkMsg(WorkContainer<?, ?> workContainer, Durat
     private boolean isOrderRestricted() {
         return options.getOrdering() != UNORDERED;
     }
+
+    // remove staled WorkContainer otherwise when the partition is reassigned, the staled messages will:
+    // 1. block the new work containers to be picked and processed
+    // 2. will cause the consumer to paused consuming new messages indefinitely
+    private boolean removeStaledWorkContainersFromShard() {
+        return this.entries.entrySet()
+                .removeIf(entry -> {
+                    WorkContainer<K, V> workContainer = entry.getValue();
+                    return pm.getPartitionState(workContainer).checkIfWorkIsStale(workContainer);
```

Comment:
This call here kills performance (see the test `lessKeysThanThreads` in `ParallelEoSStreamProcessorTest`). I wonder if there's a way for us to reduce the number of calls when cleaning up workers. We already perform the filtering, so we might want a separate job to handle stale worker removal.

@jhollandus

Reply:
Thanks for pointing out for I miss the integration test. Let me think about this one.

Reply:
@eddyv please help check the updates. I think adding another thread/job for removal of stale worker could be an overkill.
So I filter them out without break out of the loop to let normal workers could be picked up.

Reply:
Hmm... still seems to be a bit on the slow side even for just filtering. Better but still slow. I suggested the use of an extra thread, purely just so we can reduce the number of filter calls we make. Essentially, have a background thread that cleans things up every few seconds or something. Open to suggestions

Reply:
@eddyv Thanks for the comments. I put the removal of stale containers into `onPartitionsAssigned` and `onPartitionsRevoked` since:
1. these two places are the only places that the containers will be marked as stale.
2. the extra thread may introduce extra context switch overhead.

Reply:
@eddyv @jhollandus, just a kind reminder for checking the updated PR

Reply:
Thanks -- sorry, life gets in the way sometimes.

It looks like the build still times out. (btw, you can actually view the logs yourself by navigating to public ci check and clicking on view more details)

<img width="681" alt="image" src="https://github.com/confluentinc/parallel-consumer/assets/10497078/b5d0a4ac-4dc6-4f68-a307-f5e11b4652a7">

At first glance, it could be that making this call `.filter(workContainer -> !isWorkContainerStale(workContainer))` is expensive?

I wonder if it's worth it to just exclude this filter and rely on `onPartitionsAssigned` and `onPartitionsRevoked`.

Reply:
@eddyv Thanks for your valuable advice. I totally agree with you. The removal in `onPartitionsAssigned` and `onPartitionsRevoked` should be enough.
I have removed the filter part and the tests in Jenkins is passed.

Thanks for the tips for check the CI!